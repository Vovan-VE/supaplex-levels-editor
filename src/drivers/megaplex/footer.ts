import {
  FOOTER_BYTE_LENGTH,
  LevelFooter as SpLevelFooter,
} from "../supaplex/footer";
import { ILevelFooter } from "./internal";

export class LevelFooter extends SpLevelFooter implements ILevelFooter {
  #width: number;
  #demo: Uint8Array | null = null;

  constructor(width: number, data?: Uint8Array) {
    super(data?.slice(0, FOOTER_BYTE_LENGTH));
    this.#width = width;

    if (
      data &&
      data.length > FOOTER_BYTE_LENGTH &&
      data[data.length - 1] === 0xff
    ) {
      this.#demo = data.slice(FOOTER_BYTE_LENGTH + 1, data.length - 1);
    }
  }

  copy() {
    return new LevelFooter(this.width, this.getRaw()) as this;
  }

  get width() {
    return this.#width;
  }

  get length() {
    return super.length + (this.#demo ? this.#demo.length + 2 : 0);
  }

  getRaw() {
    const main = super.getRaw();
    const demo = this.#demo;
    if (demo) {
      const result = new Uint8Array(main.length + demo.length + 2);
      result.set(main, 0);
      result[main.length] = 0;
      result.set(demo, main.length + 1);
      result[main.length + 1 + demo.length] = 0xff;
      return result;
    }
    return main;
  }

  get demo() {
    return this.#demo && new Uint8Array(this.#demo);
  }

  setDemo(demo: Uint8Array | null) {
    if (demo) {
      if (
        this.#demo &&
        demo.length === this.#demo.length &&
        demo.every((v, i) => v === this.#demo![i])
      ) {
        return this;
      }
    } else {
      if (!this.#demo) {
        return this;
      }
    }
    const copy = this.copy();
    copy.#demo = demo && new Uint8Array(demo);
    return copy;
  }
}
