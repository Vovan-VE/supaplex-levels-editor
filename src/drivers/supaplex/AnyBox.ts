import { inRect, Rect } from "utils/rect";
import { ISupaplexBox } from "./internal";

export class AnyBox implements ISupaplexBox, Rect {
  readonly #width: number;
  readonly #height: number;

  readonly validateCoords =
    process.env.NODE_ENV === "production"
      ? undefined
      : (x: number, y: number) => {
          if (!inRect(x, y, this)) {
            throw new RangeError(`Cell coords (${x}, ${y}) are out of range`);
          }
        };

  constructor(width: number, height: number) {
    this.#width = width;
    this.#height = height;
  }

  get x() {
    return 0;
  }
  get y() {
    return 0;
  }

  get width() {
    return this.#width;
  }
  get height() {
    return this.#height;
  }

  get length() {
    return this.#width * this.#height;
  }

  coordsToOffset(x: number, y: number) {
    return y * this.#width + x;
  }
}
