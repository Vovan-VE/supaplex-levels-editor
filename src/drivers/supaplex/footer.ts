import { ILevelFooter } from "./internal";
import { ISupaplexSpecPort, ISupaplexSpecPortProps } from "./types";

export const FOOTER_BYTE_LENGTH = 96;

const GRAVITY_OFFSET = 4;
const TITLE_OFFSET = GRAVITY_OFFSET + 2;
export const TITLE_LENGTH = 23;
const FREEZE_ZONKS_OFFSET = TITLE_OFFSET + TITLE_LENGTH;
const INFOTRONS_NEED_OFFSET = FREEZE_ZONKS_OFFSET + 1;
const SPEC_PORT_COUNT_OFFSET = INFOTRONS_NEED_OFFSET + 1;
const SPEC_PORT_DB_OFFSET = SPEC_PORT_COUNT_OFFSET + 1;
const SPEC_PORT_ITEM_LENGTH = 6;
const SPEC_PORT_MAX_COUNT = 10;

const validateByte =
  process.env.NODE_ENV === "production"
    ? undefined
    : (byte: number) => {
        if (byte < 0 || byte > 255) {
          throw new RangeError(`Invalid byte ${byte}`);
        }
      };

const specPortRawOffset = (index: number) =>
  SPEC_PORT_DB_OFFSET + index * SPEC_PORT_ITEM_LENGTH;

const getSpecPortProps = (raw: Uint8Array): ISupaplexSpecPortProps => ({
  setsGravity: raw[2] === 1,
  setsFreezeZonks: raw[3] === 2,
  setsFreezeEnemies: raw[4] === 1,
});

export const specPortOffsetToCoords = (
  b0: number,
  b1: number,
  width: number,
): [x: number, y: number] => {
  const offset = ((b0 << 8) | b1) >> 1;
  return [offset % width, Math.floor(offset / width)];
};

export const specPortCoordsToOffset = (
  x: number,
  y: number,
  width: number,
): [b0: number, b1: number] => {
  const offset = (y * width + x) << 1;
  if (/*process.env.NODE_ENV !== "production" &&*/ offset > 0xffff) {
    throw new RangeError(
      `Spec port at coords (${x}, ${y}) with width ${width} cannot be saved`,
    );
  }
  return [(offset >> 8) & 0xff, offset & 0xff];
};

export class LevelFooter implements ILevelFooter {
  #width: number;
  #src: Uint8Array;

  constructor(width: number, data?: Uint8Array) {
    if (data) {
      if (
        process.env.NODE_ENV !== "production" &&
        data.length < FOOTER_BYTE_LENGTH
      ) {
        throw new Error(
          `Invalid buffer length ${data.length}, expected at least ${FOOTER_BYTE_LENGTH}`,
        );
      }
      data = new Uint8Array(data);
    } else {
      data = new Uint8Array(FOOTER_BYTE_LENGTH);
      data.set(
        ""
          .padEnd(TITLE_LENGTH)
          .split("")
          .map((ch) => ch.charCodeAt(0)),
        TITLE_OFFSET,
      );
    }
    this.#width = width;
    this.#src = data;
  }

  copy(): this {
    return new LevelFooter(this.#width, this.#src) as this;
  }

  get length() {
    return this.#src.length;
  }

  getRaw(width: number) {
    return new Uint8Array(this.#src);
  }

  get title() {
    return String.fromCharCode(
      ...this.#src.slice(TITLE_OFFSET, TITLE_OFFSET + TITLE_LENGTH),
    );
  }

  setTitle(title: string) {
    if (process.env.NODE_ENV !== "production" && title.length > TITLE_LENGTH) {
      throw new RangeError("Title length exceeds limit");
    }
    if (/[^\x20-\x7F]/.test(title)) {
      throw new Error("Unsupported characters found");
    }

    title = title.padEnd(TITLE_LENGTH);
    if (this.title === title) {
      return this;
    }

    const copy = this.copy();
    copy.#src.set(
      title.split("").map((ch) => ch.charCodeAt(0)),
      TITLE_OFFSET,
    );
    return copy;
  }

  get initialGravity() {
    return this.#src[GRAVITY_OFFSET] === 1;
  }
  setInitialGravity(on: boolean) {
    if (this.initialGravity === on) {
      return this;
    }
    const copy = this.copy();
    copy.#src[GRAVITY_OFFSET] = on ? 1 : 0;
    return copy;
  }

  get initialFreezeZonks() {
    return this.#src[FREEZE_ZONKS_OFFSET] === 2;
  }
  setInitialFreezeZonks(on: boolean) {
    if (this.initialFreezeZonks === on) {
      return this;
    }
    const copy = this.copy();
    copy.#src[FREEZE_ZONKS_OFFSET] = on ? 2 : 0;
    return copy;
  }

  get infotronsNeed() {
    return this.#src[INFOTRONS_NEED_OFFSET];
  }
  setInfotronsNeed(value: number) {
    validateByte?.(value);
    if (this.infotronsNeed === value) {
      return this;
    }
    const copy = this.copy();
    copy.#src[INFOTRONS_NEED_OFFSET] = value;
    return copy;
  }

  get specPortsCount() {
    return Math.min(this.#src[SPEC_PORT_COUNT_OFFSET], SPEC_PORT_MAX_COUNT);
  }

  *getSpecPorts(): Generator<ISupaplexSpecPort, void, void> {
    for (let i = 0, L = this.specPortsCount; i < L; i++) {
      const offset = specPortRawOffset(i);
      const raw = this.#src.slice(offset, offset + SPEC_PORT_ITEM_LENGTH);
      const [x, y] = specPortOffsetToCoords(raw[0], raw[1], this.#width);
      yield { x, y, ...getSpecPortProps(raw) };
    }
  }

  clearSpecPorts() {
    if (!this.specPortsCount) {
      return this;
    }
    const copy = this.copy();
    copy.#src[SPEC_PORT_COUNT_OFFSET] = 0;
    const zeroItem = new Uint8Array(SPEC_PORT_ITEM_LENGTH);
    for (let i = 0; i < SPEC_PORT_MAX_COUNT; i++) {
      copy.#src.set(zeroItem, specPortRawOffset(i));
    }
    return copy;
  }

  findSpecPort(x: number, y: number) {
    for (const port of this.getSpecPorts()) {
      if (port.x === x && port.y === y) {
        const { x: _1, y: _2, ...props } = port;
        return props;
      }
    }
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    const list = [...this.getSpecPorts()];
    let index = list.findIndex((p) => p.x === x && p.y === y);
    let newCount = this.specPortsCount;
    if (index >= 0) {
      if (!props) {
        return this;
      }
      const prev = list[index];
      if (
        props.setsGravity === prev.setsGravity &&
        props.setsFreezeZonks === prev.setsFreezeZonks &&
        props.setsFreezeEnemies === prev.setsFreezeEnemies
      ) {
        return this;
      }
    } else {
      index = this.specPortsCount;
      if (index >= SPEC_PORT_MAX_COUNT) {
        throw new RangeError("Cannot add more spec ports");
      }
      newCount++;
      props ??= {
        setsGravity: false,
        setsFreezeZonks: false,
        setsFreezeEnemies: false,
      };
    }

    const result = this.copy();
    const [b0, b1] = specPortCoordsToOffset(x, y, this.#width);
    result.#src.set(
      Uint8Array.of(
        b0,
        b1,
        props.setsGravity ? 1 : 0,
        props.setsFreezeZonks ? 2 : 0,
        props.setsFreezeEnemies ? 1 : 0,
      ),
      specPortRawOffset(index),
    );
    result.#src[SPEC_PORT_COUNT_OFFSET] = newCount;
    return result;
  }

  deleteSpecPort(x: number, y: number) {
    const list = [...this.getSpecPorts()];
    const index = list.findIndex((p) => p.x === x && p.y === y);
    if (index < 0) {
      return this;
    }
    const last = this.specPortsCount - 1;
    const copy = this.copy();
    for (let i = index; i < last; i++) {
      const from = specPortRawOffset(i + 1);
      copy.#src.set(
        this.#src.slice(from, from + SPEC_PORT_ITEM_LENGTH),
        specPortRawOffset(i),
      );
    }
    copy.#src.set(
      new Uint8Array(SPEC_PORT_ITEM_LENGTH),
      specPortRawOffset(last),
    );
    copy.#src[SPEC_PORT_COUNT_OFFSET] = this.specPortsCount - 1;
    return copy;
  }
}
