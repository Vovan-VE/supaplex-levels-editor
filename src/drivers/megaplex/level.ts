import { ILevelBody } from "../supaplex/internal";
import { LevelBody } from "../supaplex/body";
import { FOOTER_BYTE_LENGTH, TITLE_LENGTH } from "../supaplex/footer";
import { isSpecPort, TILE_SPACE } from "../supaplex/tiles";
import { ISupaplexSpecPortProps } from "../supaplex/types";
import { AnyBox } from "./box";
import { LevelFooter } from "./footer";
import { IMegaplexLevel } from "./types";

const sliceFooter = (width: number, height: number, data?: Uint8Array) => {
  if (data) {
    const bodyLength = width * height;
    const minLength = bodyLength + FOOTER_BYTE_LENGTH;
    if (process.env.NODE_ENV !== "production" && data.length < minLength) {
      throw new Error(
        `Invalid buffer length ${data.length}, expected at least ${minLength}`,
      );
    }
    return data.slice(bodyLength);
  }
};

export class MegaplexLevel extends LevelFooter implements IMegaplexLevel {
  readonly #box: AnyBox;
  #body: ILevelBody;

  constructor(width: number, height: number, data?: Uint8Array) {
    super(width, sliceFooter(width, height, data));
    this.#box = new AnyBox(width, height);
    this.#body = new LevelBody(this.#box, data?.slice(0, this.#box.length));
  }

  get length() {
    return this.#box.length + super.length;
  }

  get raw() {
    const result = new Uint8Array(this.length);
    result.set(super.getRaw(this.#box.width), this.#box.length);
    result.set(this.#body.raw, 0);
    return result;
  }

  get body() {
    return this.#body;
  }

  get width() {
    return this.#box.width;
  }

  get height() {
    return this.#box.height;
  }

  get resizable() {
    return {
      maxWidth: 0x7fff,
      maxHeight: 0x7fff,
    };
  }

  copy(): this {
    return new MegaplexLevel(
      this.#box.width,
      this.#box.height,
      this.raw,
    ) as this;
  }

  resize(width: number, height: number) {
    const origSpecPorts = [...this.getSpecPorts()];

    let src = this;
    // reducing any dimension with spec port deletion
    if (
      (width < this.#box.width && origSpecPorts.some((p) => p.x >= width)) ||
      (height < this.#box.height && origSpecPorts.some((p) => p.y >= height))
    ) {
      src = this.copy();
      // delete spec ports entry
      for (let p of origSpecPorts) {
        if (
          (width < this.#box.width && p.x >= width) ||
          (height < this.#box.height && p.y >= height)
        ) {
          src = src.setTile(p.x, p.y, TILE_SPACE);
        }
      }
    }

    const temp = new Uint8Array(width * height + src.length - src.#box.length);
    temp.set(src.getRaw(width), width * height);
    let result = new MegaplexLevel(width, height, temp) as this;

    for (let y = Math.min(height, this.#box.height); y-- > 0; ) {
      for (let x = Math.min(width, this.#box.width); x-- > 0; ) {
        result = result.setTile(x, y, src.getTile(x, y));
      }
    }

    return result;
  }

  getTile(x: number, y: number) {
    return this.#body.getTile(x, y);
  }

  setTile(x: number, y: number, value: number) {
    const prev = this.#body.getTile(x, y);
    let next = this.copy();
    next.#body = next.#body.setTile(x, y, value);
    if (isSpecPort(prev)) {
      if (!isSpecPort(value)) {
        next = next.deleteSpecPort(x, y);
      }
    } else {
      if (isSpecPort(value)) {
        next = next.setSpecPort(x, y);
      }
    }
    return next;
  }

  get maxTitleLength() {
    return TITLE_LENGTH;
  }

  findSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.findSpecPort(x, y);
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    this.#box.validateCoords?.(x, y);
    return super.setSpecPort(x, y, props);
  }

  deleteSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.deleteSpecPort(x, y);
  }
}
