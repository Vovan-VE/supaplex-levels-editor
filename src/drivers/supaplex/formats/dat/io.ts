import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

function* levelsFromBuffer(buffer: ArrayBuffer): Iterable<ISupaplexLevel> {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(buffer, i * LEVEL_BYTES_LENGTH, LEVEL_BYTES_LENGTH),
    );
  }
}

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset =>
  createLevelset(levelsFromBuffer(file));

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  const result = new Uint8Array(levelset.levelsCount * LEVEL_BYTES_LENGTH);
  let i = 0;
  for (const level of levelset.getLevels()) {
    result.set(level.raw, i * LEVEL_BYTES_LENGTH);
    i++;
  }
  return result.buffer;
};
