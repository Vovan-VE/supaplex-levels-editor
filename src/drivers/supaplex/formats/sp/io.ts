import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset => {
  if (file.byteLength < LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: less then level size");
  }

  let startOfSignature = LEVEL_BYTES_LENGTH;
  if (file.byteLength > LEVEL_BYTES_LENGTH) {
    const trailer = new Uint8Array(file, LEVEL_BYTES_LENGTH);
    const p = trailer.indexOf(0xff, 1);
    if (p > 0) {
      startOfSignature += p + 1;
    }
  }

  return createLevelset([
    createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(file, 0, startOfSignature),
    ),
  ]);
};

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  let level = levelset.getLevel(0);

  // errors covered by "support report"
  if (level.width > LEVEL_WIDTH || level.height > LEVEL_HEIGHT) {
    throw new Error(`Level is too large: ${level.width}x${level.height}`);
  }

  // warnings covered by "support report"
  if (level.width < LEVEL_WIDTH || level.height < LEVEL_HEIGHT) {
    level = level.resize({ width: LEVEL_WIDTH, height: LEVEL_HEIGHT });
  }

  return level.raw.buffer;
};
