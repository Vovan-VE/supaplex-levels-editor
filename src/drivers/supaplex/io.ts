import { ISupaplexLevelset, ISupaplexReader, ISupaplexWriter } from "./types";
import { LEVEL_BYTES_LENGTH, SupaplexLevel } from "./level";
import { SupaplexLevelset } from "./levelset";

function* levelsFromBuffer(buffer: ArrayBuffer) {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield SupaplexLevel.fromArrayBuffer(buffer, i * LEVEL_BYTES_LENGTH);
  }
}

function* levelsToRaw(levelset: ISupaplexLevelset) {
  for (const level of levelset.getLevels()) {
    yield level.raw;
  }
}

export const readLevelset = async (file: Blob): Promise<ISupaplexLevelset> =>
  new SupaplexLevelset(levelsFromBuffer(await file.arrayBuffer()));

export const writeLevelset = async (
  levelset: ISupaplexLevelset,
): Promise<Blob> => new Blob([...levelsToRaw(levelset)]);

export const reader: ISupaplexReader = {
  readLevelset,
};

export const writer: ISupaplexWriter = {
  writeLevelset,
};
