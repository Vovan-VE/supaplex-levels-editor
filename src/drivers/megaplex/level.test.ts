import { createLevel, MegaplexLevel } from "./level";
import { FOOTER_BYTE_LENGTH, TITLE_LENGTH } from "../supaplex/footer";
import { dumpLevel, readExampleFile } from "./helpers.dev";
import {
  TILE_HARDWARE,
  TILE_INFOTRON,
  TILE_SP_PORT_D,
  TILE_SP_PORT_R,
  TILE_SP_PORT_U,
  TILE_ZONK,
} from "../supaplex/tiles-id";
import {
  ISupaplexSpecPort,
  ISupaplexSpecPortProps,
} from "../supaplex/internal";
import { reader } from "./io";

describe("level", () => {
  const testFooter = Uint8Array.of(
    ...[0, 0, 0, 0],
    // G
    1,
    ...[0],
    // title
    ..."-- Lorem ipsum --"
      .padEnd(23)
      .split("")
      .map((c) => c.charCodeAt(0)),
    // freeze zonks
    2,
    // infotrons need
    42,
    // spec ports count
    1,
    // spec port db
    ..."\x00\x02\x01\x02\x01\x00"
      .padEnd(10 * 6, "\x00")
      .split("")
      .map((c) => c.charCodeAt(0)),
    ...[0, 0, 0, 0],
  );
  expect(testFooter.length).toBe(96);

  const testLevelData = Uint8Array.of(
    0,
    TILE_SP_PORT_U,
    4,
    1,
    3,
    5,
    ...testFooter,
  );

  describe("constructor", () => {
    it("no data", () => {
      const level = createLevel(3, 2);

      expect(level.width).toBe(3);
      expect(level.height).toBe(2);
      expect(level.length).toBe(6 + FOOTER_BYTE_LENGTH);

      expect(level.getTile(0, 0)).toEqual(0);
      expect(level.getTile(2, 1)).toEqual(0);
      expect(level.title).toBe("".padEnd(TITLE_LENGTH));
    });

    it("data", () => {
      const level = createLevel(3, 2, testLevelData);

      expect(dumpLevel(level)).toMatchSnapshot();
    });

    it("throw", () => {
      expect(() => createLevel(3, 2, new Uint8Array(42))).toThrow(
        new Error(
          `Invalid buffer length 42, expected at least ${
            6 + FOOTER_BYTE_LENGTH
          }`,
        ),
      );
    });
  });

  it("raw", () => {
    const level = createLevel(3, 2, testLevelData);
    expect(level.raw).toEqual(testLevelData);
    expect(level.length).toEqual(testLevelData.length);

    const demo = Uint8Array.of(10, 20, 30, 40, 50, 60, 70);
    const next = level.setDemoSeed({ lo: 0x12, hi: 0x34 }).setDemo(demo);
    expect(next.length).toEqual(testLevelData.length + 7 + 2);
    expect(next.length).toEqual(6 + 96 + 7 + 2);

    const modFooter = new Uint8Array(testFooter);
    modFooter[94] = 0x12;
    modFooter[95] = 0x34;
    expect(next.raw).toEqual(
      Uint8Array.of(
        0,
        TILE_SP_PORT_U,
        4,
        1,
        3,
        5,
        ...modFooter,
        0,
        ...demo,
        0xff,
      ),
    );
  });

  it("copy", () => {
    const a = createLevel(3, 2)
      .setTitle("First level title")
      .setTile(1, 0, 6) as MegaplexLevel;

    let b = a.copy();
    expect(dumpLevel(b)).toEqual(dumpLevel(a));

    b = b.setTitle("Copy level title").setTile(1, 0, 1);
    let dump = dumpLevel(b);
    expect(dump).not.toEqual(dumpLevel(a));
    expect(dump).toMatchSnapshot();
  });

  describe("resize", () => {
    it("simple", () => {
      const a = createLevel(3, 2, testLevelData);

      let b = a.resize(5, 4);
      expect(dumpLevel(b)).toMatchSnapshot();

      b = b
        .setTile(3, 0, TILE_INFOTRON)
        .setTile(4, 0, TILE_SP_PORT_R)
        .setTile(0, 2, TILE_ZONK)
        .setTile(0, 3, TILE_SP_PORT_D)
        .setSpecPort(4, 0, {
          setsGravity: true,
          setsFreezeZonks: false,
          setsFreezeEnemies: true,
        })
        .setSpecPort(0, 3, {
          setsGravity: false,
          setsFreezeZonks: true,
          setsFreezeEnemies: true,
        });
      expect(dumpLevel(b)).toMatchSnapshot();

      const c = b.resize(3, 2);
      expect(dumpLevel(c)).toEqual(dumpLevel(a));
    });
  });

  it("getTile", () => {
    const level = createLevel(3, 2, testLevelData);

    expect(level.getTile(0, 0)).toEqual(0);
    expect(level.getTile(1, 0)).toEqual(TILE_SP_PORT_U);
    expect(level.getTile(2, 0)).toEqual(4);
    expect(level.getTile(0, 1)).toEqual(1);
    expect(level.getTile(1, 1)).toEqual(3);
    expect(level.getTile(2, 1)).toEqual(5);
  });

  describe("setTile", () => {
    it("usual", () => {
      const level = createLevel(3, 2, testLevelData);
      const copy = level.setTile(2, 0, 7);
      expect(level.getTile(2, 0)).toBe(4);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(2, 0)).toBe(7);
      expect(copy.specPortsCount).toBe(1);
    });

    it("noop", () => {
      const level = createLevel(3, 2, testLevelData);
      expect(level.setTile(2, 0, level.getTile(2, 0))).toBe(level);
    });

    it("add spec port", () => {
      const level = createLevel(3, 2, testLevelData);
      const copy = level.setTile(2, 0, TILE_SP_PORT_U);
      expect(level.getTile(2, 0)).toBe(4);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(2, 0)).toBe(TILE_SP_PORT_U);
      expect(copy.specPortsCount).toBe(2);
      expect([...copy.getSpecPorts()]).toEqual<ISupaplexSpecPort[]>([
        {
          x: 1,
          y: 0,
          setsGravity: true,
          setsFreezeZonks: true,
          setsFreezeEnemies: true,
        },
        {
          x: 2,
          y: 0,
          setsGravity: false,
          setsFreezeZonks: false,
          setsFreezeEnemies: false,
        },
      ]);
    });

    it("keep spec port", () => {
      const level = createLevel(3, 2, testLevelData);
      const copy = level.setTile(1, 0, TILE_SP_PORT_R);
      expect(level.getTile(1, 0)).toBe(TILE_SP_PORT_U);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(1, 0)).toBe(TILE_SP_PORT_R);
      expect(copy.specPortsCount).toBe(1);
      expect([...copy.getSpecPorts()]).toEqual<ISupaplexSpecPort[]>([
        {
          x: 1,
          y: 0,
          setsGravity: true,
          setsFreezeZonks: true,
          setsFreezeEnemies: true,
        },
      ]);
    });

    it("remove spec port", () => {
      const level = createLevel(3, 2, testLevelData);
      const copy = level.setTile(1, 0, TILE_ZONK);
      expect(copy.getTile(1, 0)).toBe(TILE_ZONK);
      expect(copy.specPortsCount).toBe(0);
      expect([...copy.getSpecPorts()]).toEqual([]);
    });
  });

  it("findSpecPort", () => {
    const level = createLevel(3, 2, testLevelData);
    expect(level.findSpecPort(1, 0)).toEqual<ISupaplexSpecPortProps>({
      setsGravity: true,
      setsFreezeZonks: true,
      setsFreezeEnemies: true,
    });
  });

  describe("tilesRenderStream", () => {
    it("huge maze", async () => {
      const data = await readExampleFile("HUDEMAZ1.mpx");
      const level = reader.readLevelset(data.buffer).getLevel(0);

      const a = [...level.tilesRenderStream(0, 0, 202, 202)];
      expect(a.length).toBeLessThan(202 * 202);
      expect(a[0]).toEqual([0, 0, 202, TILE_HARDWARE]);
      expect(a[a.length - 1]).toEqual([0, 201, 202, TILE_HARDWARE]);
    });
  });
});
