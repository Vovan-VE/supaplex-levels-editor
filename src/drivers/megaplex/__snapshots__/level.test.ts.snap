// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`level constructor data 1`] = `
Object {
  "cells": Array [
    " ∩&",
    "o@▢",
  ],
  "demo": null,
  "height": 2,
  "infotronsNeed": 42,
  "initialFreezeZonks": true,
  "initialGravity": true,
  "maxTitleLength": 23,
  "resizable": Object {
    "maxHeight": 32767,
    "maxWidth": 32767,
  },
  "specialPorts": Array [
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": true,
      "setsGravity": true,
      "x": 1,
      "y": 0,
    },
  ],
  "title": "-- Lorem ipsum --      ",
  "width": 3,
}
`;

exports[`level resize simple 1`] = `
Object {
  "cells": Array [
    " ∩&  ",
    "o@▢  ",
    "     ",
    "     ",
  ],
  "demo": null,
  "height": 4,
  "infotronsNeed": 42,
  "initialFreezeZonks": true,
  "initialGravity": true,
  "maxTitleLength": 23,
  "resizable": Object {
    "maxHeight": 32767,
    "maxWidth": 32767,
  },
  "specialPorts": Array [
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": true,
      "setsGravity": true,
      "x": 1,
      "y": 0,
    },
  ],
  "title": "-- Lorem ipsum --      ",
  "width": 5,
}
`;

exports[`level resize simple 2`] = `
Object {
  "cells": Array [
    " ∩&&⊃",
    "o@▢  ",
    "o    ",
    "∪    ",
  ],
  "demo": null,
  "height": 4,
  "infotronsNeed": 42,
  "initialFreezeZonks": true,
  "initialGravity": true,
  "maxTitleLength": 23,
  "resizable": Object {
    "maxHeight": 32767,
    "maxWidth": 32767,
  },
  "specialPorts": Array [
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": true,
      "setsGravity": true,
      "x": 1,
      "y": 0,
    },
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": false,
      "setsGravity": true,
      "x": 4,
      "y": 0,
    },
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": true,
      "setsGravity": false,
      "x": 0,
      "y": 3,
    },
  ],
  "title": "-- Lorem ipsum --      ",
  "width": 5,
}
`;
